planet/clients/orders.py

```diff
@@ -337,71 +340,90 @@ async def download_order(
+        specified state never being reached (i.e. if the order completes in
+        a 'partial' state but only the 'success' state was specified).
 
         Example:
             ```python
             from planet.reporting import StateBar
 
             with StateBar() as bar:
-                await poll(order_id, report=bar.update)
+                await wait(order_id, report=bar.update_state)
             ```
 
         Parameters:
             order_id: The ID of the order
-            state: State to poll until. If multiple, use list. Defaults to
-                any completed state.
-            wait: Time (in seconds) between polls.
-            report: Callback function for progress updates. Invoked with
-                keyword arguments `state` (poll state) and `logger`
-                (callback for logging progress bar status). Recommended
-                value is `reporting.StateBar.update`.
+            states: Order states that will end polling. Defaults to all
+                completed states.
+            delay: Time (in seconds) between polls.
+            max_attempts: Maximum number of polls. Set to zero for no limit.
+            report: Callback function for reporting progress updates.
 
         Returns
-            Completed state of the order.
+            State of the order.
 
         Raises:
             planet.exceptions.APIException: On API error.
-            OrdersClientException: If order_id is not valid or state is not
-                supported.
+            planet.exceptions.ValueError: If order_id or one or more of the
+                states are not valid.
+            planet.exceptions.MaxAttemptsError: If the maximum number of
+                attempts is reached before one of the specified states is
+                reached.
         """
-        if state:
-            if state not in ORDERS_STATES:
-                raise OrdersClientException(
-                    f'{state} should be one of'
-                    f'{ORDERS_STATES}')
-            states = [state]
-        else:
-            states = ORDERS_STATES_COMPLETE
-
-        completed = False
-        while not completed:
+        invalid_states = set(states) - ORDERS_STATES_ALL
+        if invalid_states:
+            raise exceptions.ValueError(
+                f'{invalid_states} are not valid states. '
+                f'Valid states are {ORDERS_STATES_ALL}')
+
+        num_attempts = 0
+        done = False
+        while not done:
             t = time.time()
 
             order = await self.get_order(order_id)
             state = order.state
 
+            LOGGER.debug(state)
+
             if report:
-                report(state=order.state)
-            else:
-                LOGGER.debug(state)
+                report(order.state)
 
-            completed = state in states
-            if not completed:
-                sleep_time = max(wait-(time.time()-t), 0)
+            done = state in states
+            if not done:
```

Comment:
Two different kinds of conditionals testing `done` is a bit of a red flag for me. I think this code would be easier to understand if it used one `while i < max_attempts` and used `break` to get out early.

Reply:
that makes sense. I'll give it a go.